crates/wasmtime/src/component/types.rs

```diff
@@ -0,0 +1,1025 @@
+                            .check(&value)
+                            .context("type mismatch for err case of expected")?;
+                        value
+                    }
+                }),
+            }))
+        } else {
+            Err(anyhow!("cannot make expected from {} type", self.desc()))
+        }
+    }
+
+    /// Instantiate this type (which must be a `Type::Flags`) with the specified flag `names`.
+    pub fn new_flags(&self, names: &[&str]) -> Result<Val> {
+        if let Type::Flags(handle) = self {
+            let map = handle
+                .names()
+                .enumerate()
+                .map(|(index, name)| (name, index))
+                .collect::<HashMap<_, _>>();
+
+            let mut values = vec![0_u32; values::u32_count_for_flag_count(handle.names().len())];
+
+            for name in names {
+                let index = map
+                    .get(name)
+                    .ok_or_else(|| anyhow!("unknown flag: {name}"))?;
+                values[index / 32] |= 1 << (index % 32);
+            }
+
+            Ok(Val::Flags(Flags {
+                ty: handle.clone(),
+                count: u32::try_from(map.len())?,
+                value: values.into(),
+            }))
+        } else {
+            Err(anyhow!("cannot make flags from {} type", self.desc()))
+        }
+    }
+
+    /// Retrieve any types nested within this type.
+    ///
+    /// For example, this will return the field types in order of declaration for a record type.  It will return
+    /// the element type for a list type and nothing for non-composite types like bool, u8, and string.
+    pub fn nested(&self) -> Box<[Type]> {
+        match self {
+            Type::Unit
+            | Type::Bool
+            | Type::S8
+            | Type::U8
+            | Type::S16
+            | Type::U16
+            | Type::S32
+            | Type::U32
+            | Type::S64
+            | Type::U64
+            | Type::Float32
+            | Type::Float64
+            | Type::Char
+            | Type::String
+            | Type::Enum(_)
+            | Type::Flags(_) => Box::new([]),
+
+            Type::List(handle) => Box::new([handle.ty()]),
+
+            Type::Record(handle) => handle.fields().map(|field| field.ty).collect(),
+
+            Type::Tuple(handle) => handle.types().collect(),
+
+            Type::Variant(handle) => handle.cases().map(|case| case.ty).collect(),
+
+            Type::Union(handle) => handle.types().collect(),
+
+            Type::Option(handle) => Box::new([handle.ty()]),
+
+            Type::Expected(handle) => Box::new([handle.ok(), handle.err()]),
+        }
+    }
+
+    pub(crate) fn check(&self, value: &Val) -> Result<()> {
+        if self == &value.ty() {
+            Ok(())
+        } else {
+            Err(anyhow!(
+                "type mismatch: expected {}, got {}",
+                self.desc(),
+                value.ty().desc()
+            ))
+        }
```

Comment:
This doesn't necessarily have to be handled here, but I think this typecheck will eventually want to be more recursive for better error mesages. Right now I think you could get `expected record, got record` which doesn't describe which record fields mismatch or similar.

Reply:
Any e.g. record field mismatch would have been caught already in `Type::new_record` before we got this far.  The only time you'd see `expected record, got record` would be if you used `Val`s from one component in another, or if your component had two identical `record`s (i.e. same field names and types) and you mixed up which one you created the `Val` for.  Either way, more recursion wouldn't necessarily help.  This is somewhat analogous to when `rustc` reports a type mismatch between `foo::Bar` and `foo::Bar` because you're using two different versions of the `foo` crate.  It's even worse here since types don't really have names in the component model.

But yes, I agree we should give a friendlier diagnostic here, e.g. hinting based on whether `Handle::index` or `Handle::types` are different, or both.

Reply:
I've improved the error message so you'll never get `expected record, got record`.  I attempted to go much further than that and recursively compare the types, but reusing `typed::typecheck_tuple` etc. turned out to be difficult in a dynamic context, and I didn't want to duplicate them either.

Also, I'm having trouble finding guidance in https://github.com/WebAssembly/component-model about when interface types are to be considered equal and when not, so I'm not sure what the correct logic is anyway.  Are types considered equal if they are structurally identical?

Reply:
I don't think we'll get a ton of guidance from the component model itself, but for now from Wasmtime's perspective given how trampolines and lifting/lowering/etc are all implemented we require exact structural equality for the embedder API. I think the intention is that subtyping would eventually be used but we don't have a great means of implementing that right now.